# 01 Importing Libraries 

In [1]:
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [2]:
path = r'/Users/woodoooo/Desktop/Instacart Basket Analysis'

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', '4.7_ords_prods_merge_Part 2.pkl'))

In [3]:
#departments data 
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'), index_col = False)

In [4]:
df_dep

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [5]:
df_dep = df_dep.rename(columns={'Unnamed: 0': 'department_id'})

In [6]:
# Change data types for multiple columns
df_ords_prods_merge = df_ords_prods_merge.astype({
    'order_id': 'int32',
    'user_id': 'int32',
    'order_number': 'int8',
    'order_day_of_week': 'int8',
    'order_hour_of_day': 'int8',
    'product_id': 'int32',
    'add_to_cart_order': 'int32',
    'reordered': 'int8',
    'days_since_prior_order': 'float32',
    'aisle_id': 'int32',
    'department_id': 'int32',
    'prices': 'float32'
    })

In [7]:
df_ords_prods_merge.shape

(32404859, 19)

In [8]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   
5   2398795        1             2                  3                  7   
6   2398795        1             2                  3                  7   
7   2398795        1             2                  3                  7   
8   2398795        1             2                  3                  7   
9   2398795        1             2                  3                  7   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   
5                    15.0         False         196                  1   
6                    15.0         False       10258                  2   
7                    15.0         False       12427                  3   
8                    15.0         False       13176                  4   
9                    15.0         False       26088                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   
5          1                                     Soda        77   
6          0                               Pistachios       117   
7          1                      Original Beef Jerky        23   
8          0                   Bag of Organic Bananas        24   
9          1               Aged White Cheddar Popcorn        23   

   department_id  prices match price_range_loc     Busiest day  \
0              7     9.0  both       Mid-range  Regularly busy   
1             16    12.5  both       Mid-range  Regularly busy   
2             19     4.4  both       Low-range  Regularly busy   
3             19     4.7  both       Low-range  Regularly busy   
4             17     1.0  both       Low-range  Regularly busy   
5              7     9.0  both       Mid-range  Regularly busy   
6             19     3.0  both       Low-range  Regularly busy   
7             19     4.4  both       Low-range  Regularly busy   
8              4    10.3  both       Mid-range  Regularly busy   
9             19     4.7  both       Low-range  Regularly busy   

         Busiest_days busiest_period_of_day  
0  Regularly busy day        Average orders  
1  Regularly busy day        Average orders  
2  Regularly busy day        Average orders  
3  Regularly busy day        Average orders  
4  Regularly busy day        Average orders  
5         Slowest day         Fewest orders  
6         Slowest day         Fewest orders  
7         Slowest day         Fewest orders  
8         Slowest day         Fewest orders  
9         Slowest day         Fewest orders

# 03 Grouping (Aggregated Mean Calculation for the Entire Dataframe)

In [9]:
df_ords_prods_merge.groupby('product_name')

In [10]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [11]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': 'mean'}).sort_values('order_number', ascending=False)

order_number
department_id              
21                22.902379
10                20.197148
18                19.310397
4                 17.811403
16                17.665606
2                 17.277920
7                 17.225802
19                17.177343
3                 17.170395
14                16.773669
13                16.583536
20                16.473447
6                 16.439806
11                16.170638
15                16.165037
9                 15.895474
12                15.887671
17                15.694469
1                 15.457838
8                 15.340650
5                 15.215751

In [12]:
dept_mean = df_ords_prods_merge.groupby('department_id') \
    .agg({'order_number': 'mean'}) \
    .sort_values('order_number', ascending=False) \
    .reset_index()

In [13]:
# merge with department names
dept_mean_named = dept_mean.merge(df_dep, on='department_id', how='inner')

In [14]:
dept_mean_named.sort_values('order_number', ascending=False)

department_id  order_number       department
0              21     22.902379          missing
1              10     20.197148             bulk
2              18     19.310397           babies
3               4     17.811403          produce
4              16     17.665606       dairy eggs
5               2     17.277920            other
6               7     17.225802        beverages
7              19     17.177343           snacks
8               3     17.170395           bakery
9              14     16.773669        breakfast
10             13     16.583536           pantry
11             20     16.473447             deli
12              6     16.439806    international
13             11     16.170638    personal care
14             15     16.165037     canned goods
15              9     15.895474  dry goods pasta
16             12     15.887671     meat seafood
17             17     15.694469        household
18              1     15.457838           frozen
19              8     15.340650             pets
20              5     15.215751          alcohol

### Analysis of Aggregated Mean Order Numbers

**Ranking is mostly consistent.**  
  Departments like **21 (missing)**, **10 (bulk)**, and **18 (babies)** remain at the top in both the sample and full dataset. This suggests the sample did capture the general pattern of customer behavior.

**Mean values changed slightly.**  
  For example, Department 21 had an average of **21.96** in the sample, but **22.90** in the full dataset. These small shifts reflect the fact that the full dataset provides a more accurate average.

**Some departments shifted positions.**  
  - In the sample, Department 16 (dairy eggs) ranked higher (18.01) but in the full dataset it dropped slightly (17.66).  
  - Minor swaps like this show how smaller samples can slightly skew rankings.

 **Conclusion:**  
  The **sample** is useful for testing and quick exploration, but the **full dataframe** gives more reliable results for business decisions — especially when comparing departments or presenting final insights.

# 04 Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions. 

In [15]:
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [16]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [17]:
df_ords_prods_merge.loc[
   (df_ords_prods_merge['max_order'] <= 40) & 
   (df_ords_prods_merge['max_order'] > 10), 
'loyalty_flag'] = 'Regular customer'

In [18]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [19]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   
5   2398795        1             2                  3                  7   
6   2398795        1             2                  3                  7   
7   2398795        1             2                  3                  7   
8   2398795        1             2                  3                  7   
9   2398795        1             2                  3                  7   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   
5                    15.0         False         196                  1   
6                    15.0         False       10258                  2   
7                    15.0         False       12427                  3   
8                    15.0         False       13176                  4   
9                    15.0         False       26088                  5   

   reordered  ... aisle_id  department_id  prices  match price_range_loc  \
0          0  ...       77              7     9.0   both       Mid-range   
1          0  ...       91             16    12.5   both       Mid-range   
2          0  ...       23             19     4.4   both       Low-range   
3          0  ...       23             19     4.7   both       Low-range   
4          0  ...       54             17     1.0   both       Low-range   
5          1  ...       77              7     9.0   both       Mid-range   
6          0  ...      117             19     3.0   both       Low-range   
7          1  ...       23             19     4.4   both       Low-range   
8          0  ...       24              4    10.3   both       Mid-range   
9          1  ...       23             19     4.7   both       Low-range   

      Busiest day        Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy day        Average orders        10   
1  Regularly busy  Regularly busy day        Average orders        10   
2  Regularly busy  Regularly busy day        Average orders        10   
3  Regularly busy  Regularly busy day        Average orders        10   
4  Regularly busy  Regularly busy day        Average orders        10   
5  Regularly busy         Slowest day         Fewest orders        10   
6  Regularly busy         Slowest day         Fewest orders        10   
7  Regularly busy         Slowest day         Fewest orders        10   
8  Regularly busy         Slowest day         Fewest orders        10   
9  Regularly busy         Slowest day         Fewest orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  
5  New customer  
6  New customer  
7  New customer  
8  New customer  
9  New customer  

[10 rows x 21 columns]

In [57]:
df_ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

In [21]:
df_ords_prods_merge.shape

(32404859, 21)

# 05 Business questions
### The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [22]:
# Group by the loyalty flag and calculate price statistics
loyalty_price_stats = df_ords_prods_merge.groupby('loyalty_flag')['prices'].describe()
loyalty_price_stats

count      mean       std  min  25%  50%   75%   max
loyalty_flag                                                               
Loyal customer    10284093.0  7.773498  4.178565  1.0  4.2  7.4  11.2  25.0
New customer       6243990.0  7.801071  4.261004  1.0  4.2  7.4  11.3  25.0
Regular customer  15876776.0  7.798141  4.204508  1.0  4.2  7.4  11.3  25.0

**Mean price per product:**

- Loyal customers: **$7.77**
  
- New customers: **$7.80**
  
- Regular customers: **$7.79**

- The median (50%) is the same across all groups: **$7.4**

Interpretation:
There is **almost no difference** in average product prices between the three loyalty groups. Loyal customers aren’t buying significantly more expensive or cheaper products compared to new or regular customers. This means marketing campaigns might not need to differentiate by product pricing, but should instead focus on **other behavioral factors** (e.g., frequency of purchase, product categories).


### The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria: 
### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” 
### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [23]:
user_avg_price = df_ords_prods_merge.groupby('user_id')['prices'].mean()

In [24]:
user_avg_price

user_id
1         6.367797
2         7.515897
3         8.197727
4         8.205555
5         9.189189
            ...   
206205    8.909375
206206    7.646667
206207    7.313453
206208    8.366618
206209    7.058915
Name: prices, Length: 206209, dtype: float32

In [25]:
df_ords_prods_merge['avg_price'] = df_ords_prods_merge['user_id'].map(user_avg_price)

In [26]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... department_id  prices  match  price_range_loc  \
0          0  ...             7     9.0   both        Mid-range   
1          0  ...            16    12.5   both        Mid-range   
2          0  ...            19     4.4   both        Low-range   
3          0  ...            19     4.7   both        Low-range   
4          0  ...            17     1.0   both        Low-range   

      Busiest day        Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy day        Average orders        10   
1  Regularly busy  Regularly busy day        Average orders        10   
2  Regularly busy  Regularly busy day        Average orders        10   
3  Regularly busy  Regularly busy day        Average orders        10   
4  Regularly busy  Regularly busy day        Average orders        10   

   loyalty_flag  avg_price  
0  New customer   6.367797  
1  New customer   6.367797  
2  New customer   6.367797  
3  New customer   6.367797  
4  New customer   6.367797  

[5 rows x 22 columns]

In [27]:
df_ords_prods_merge.shape

(32404859, 22)

In [28]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['avg_price'] >= 10, 'avg_spending_flag'] = 'High spender'

In [29]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['avg_price'] < 10, 'avg_spending_flag'] = 'Low spender'

In [30]:
df_ords_prods_merge.head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2539329        1             1                  2                  8   
2    2539329        1             1                  2                  8   
3    2539329        1             1                  2                  8   
4    2539329        1             1                  2                  8   
5    2398795        1             2                  3                  7   
6    2398795        1             2                  3                  7   
7    2398795        1             2                  3                  7   
8    2398795        1             2                  3                  7   
9    2398795        1             2                  3                  7   
10   2398795        1             2                  3                  7   
11    473747        1             3                  3                 12   
12    473747        1             3                  3                 12   
13    473747        1             3                  3                 12   
14    473747        1             3                  3                 12   

    days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                      NaN          True         196                  1   
1                      NaN          True       14084                  2   
2                      NaN          True       12427                  3   
3                      NaN          True       26088                  4   
4                      NaN          True       26405                  5   
5                     15.0         False         196                  1   
6                     15.0         False       10258                  2   
7                     15.0         False       12427                  3   
8                     15.0         False       13176                  4   
9                     15.0         False       26088                  5   
10                    15.0         False       13032                  6   
11                    21.0         False         196                  1   
12                    21.0         False       12427                  2   
13                    21.0         False       10258                  3   
14                    21.0         False       25133                  4   

    reordered  ... prices  match  price_range_loc     Busiest day  \
0           0  ...    9.0   both        Mid-range  Regularly busy   
1           0  ...   12.5   both        Mid-range  Regularly busy   
2           0  ...    4.4   both        Low-range  Regularly busy   
3           0  ...    4.7   both        Low-range  Regularly busy   
4           0  ...    1.0   both        Low-range  Regularly busy   
5           1  ...    9.0   both        Mid-range  Regularly busy   
6           0  ...    3.0   both        Low-range  Regularly busy   
7           1  ...    4.4   both        Low-range  Regularly busy   
8           0  ...   10.3   both        Mid-range  Regularly busy   
9           1  ...    4.7   both        Low-range  Regularly busy   
10          0  ...    4.0   both        Low-range  Regularly busy   
11          1  ...    9.0   both        Mid-range  Regularly busy   
12          1  ...    4.4   both        Low-range  Regularly busy   
13          1  ...    3.0   both        Low-range  Regularly busy   
14          0  ...    8.6   both        Mid-range  Regularly busy   

          Busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy day        Average orders        10  New customer   
1   Regularly busy day        Average orders        10  New customer   
2   Regularly busy day        Average orders        10  New customer   
3   Regularly busy day        Average orders        10  New customer   
4   Regularly busy day        Average orders        10  New customer   
5          Slowest day         Fe

In [31]:
df_ords_prods_merge.shape

(32404859, 23)

In [32]:
df_ords_prods_merge['avg_spending_flag'].value_counts(dropna = False)

avg_spending_flag
Low spender     32285150
High spender      119709
Name: count, dtype: int64

In [33]:
# Group by the 'avg_spending_flag' flag and calculate statistics
avg_user_spending_stats = df_ords_prods_merge.groupby('avg_spending_flag')['prices'].describe()
avg_user_spending_stats

count       mean       std  min  25%   50%   75%   max
avg_spending_flag                                                             
High spender         119709.0  10.578568  4.739679  1.0  7.3  11.2  13.5  25.0
Low spender        32285150.0   7.780550  4.089216  1.0  4.2   7.4  11.2  25.0

Key Findings:
- **High spenders consistently purchase items with higher average prices.**
- **Low spenders focus more on lower-priced products**, with a mean under $8.

Conclusion:
The spending flag effectively separates customers into **two distinct price behavior groups**.  
This segmentation can help the marketing team design **different campaigns** — for example:
- Promoting **premium products** to high spenders.
- Offering **discounts or entry-level deals** to low spenders to encourage more frequent or higher-value purchases.
This keeps the analysis clear and business‑oriented, and it directly ties your stats back to the marketing strategy.


### In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [34]:
# We should use here Median values instead of the average (mean) because:

# Median is less sensitive to extreme values
# Some customers might have very long breaks (e.g., didn’t order for 150 days) or odd one-time gaps.
# These outliers would inflate the mean and make a customer look less frequent than they really are.

In [35]:
df_ords_prods_merge['user_median_order_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [36]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... match  price_range_loc     Busiest day        Busiest_days  \
0          0  ...  both        Mid-range  Regularly busy  Regularly busy day   
1          0  ...  both        Mid-range  Regularly busy  Regularly busy day   
2          0  ...  both        Low-range  Regularly busy  Regularly busy day   
3          0  ...  both        Low-range  Regularly busy  Regularly busy day   
4          0  ...  both        Low-range  Regularly busy  Regularly busy day   

  busiest_period_of_day max_order  loyalty_flag avg_price avg_spending_flag  \
0        Average orders        10  New customer  6.367797       Low spender   
1        Average orders        10  New customer  6.367797       Low spender   
2        Average orders        10  New customer  6.367797       Low spender   
3        Average orders        10  New customer  6.367797       Low spender   
4        Average orders        10  New customer  6.367797       Low spender   

   user_median_order_frequency  
0                         20.5  
1                         20.5  
2                         20.5  
3                         20.5  
4                         20.5  

[5 rows x 24 columns]

In [38]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['user_median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [39]:
df_ords_prods_merge.loc[
   (df_ords_prods_merge['user_median_order_frequency'] <=20) &
   (df_ords_prods_merge['user_median_order_frequency'] >10),
   'order_frequency_flag'] = 'Regular customer'

In [40]:
df_ords_prods_merge.loc[
   df_ords_prods_merge['user_median_order_frequency'] <=10, 'order_frequency_flag'] = 'Frequent customer'

In [41]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... price_range_loc     Busiest day        Busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regularly busy day   
1          0  ...       Mid-range  Regularly busy  Regularly busy day   
2          0  ...       Low-range  Regularly busy  Regularly busy day   
3          0  ...       Low-range  Regularly busy  Regularly busy day   
4          0  ...       Low-range  Regularly busy  Regularly busy day   

   busiest_period_of_day max_order  loyalty_flag avg_price avg_spending_flag  \
0         Average orders        10  New customer  6.367797       Low spender   
1         Average orders        10  New customer  6.367797       Low spender   
2         Average orders        10  New customer  6.367797       Low spender   
3         Average orders        10  New customer  6.367797       Low spender   
4         Average orders        10  New customer  6.367797       Low spender   

  user_median_order_frequency   order_frequency_flag  
0                        20.5  Non-frequent customer  
1                        20.5  Non-frequent customer  
2                        20.5  Non-frequent customer  
3                        20.5  Non-frequent customer  
4                        20.5  Non-frequent customer  

[5 rows x 25 columns]

In [42]:
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [43]:
df_ords_prods_merge.shape

(32404859, 25)

# Additional check -- Top products by High- and Low Spender

In [44]:
# Group by spending flag and product

products_by_spender = (
    df_ords_prods_merge
    .groupby(['avg_spending_flag', 'department_id'])['product_id']
    .count()
    .reset_index()
    .sort_values(['avg_spending_flag', 'product_id'], ascending=[True, False])
)

In [45]:
# 1️⃣ Filter only High spenders
high_spender_products = products_by_spender[products_by_spender['avg_spending_flag'] == 'High spender']

# 2️⃣ Sort by product_id in descending order
high_spender_products = high_spender_products.sort_values('product_id', ascending=False)

# 3️⃣ Merge with department dictionary to get names
high_spender_products = high_spender_products.merge(df_dep.rename(columns={'Unnamed: 0': 'department_id'}),
                                                    on='department_id', how='left')

high_spender_products

avg_spending_flag  department_id  product_id       department
0       High spender              4       31652          produce
1       High spender             16       23112       dairy eggs
2       High spender              7       12598        beverages
3       High spender             12       10221     meat seafood
4       High spender              1        7871           frozen
5       High spender             13        7340           pantry
6       High spender              3        3952           bakery
7       High spender             19        3769           snacks
8       High spender             20        3326             deli
9       High spender             17        2939        household
10      High spender             15        2857     canned goods
11      High spender              9        2483  dry goods pasta
12      High spender             14        2149        breakfast
13      High spender             11        1955    personal care
14      High spender              5        1432          alcohol
15      High spender              6         839    international
16      High spender             18         434           babies
17      High spender              8         387             pets
18      High spender             21         245          missing
19      High spender              2          80            other
20      High spender             10          68             bulk

In [46]:
# 1️⃣ Filter only Low spenders
low_spender_products = products_by_spender[products_by_spender['avg_spending_flag'] == 'Low spender']

# 2️⃣ Sort by product_id in descending order
low_spender_products = low_spender_products.sort_values('product_id', ascending=False)

# 3️⃣ Merge with department dictionary to get names
low_spender_products = low_spender_products.merge(df_dep.rename(columns={'Unnamed: 0': 'department_id'}),
                                                    on='department_id', how='left')

low_spender_products

avg_spending_flag  department_id  product_id       department
0        Low spender              4     9447639          produce
1        Low spender             16     5375635       dairy eggs
2        Low spender             19     2883781           snacks
3        Low spender              7     2675525        beverages
4        Low spender              1     2226872           frozen
5        Low spender             13     1868029           pantry
6        Low spender              3     1168476           bakery
7        Low spender             15     1065201     canned goods
8        Low spender             20     1047923             deli
9        Low spender              9      864144  dry goods pasta
10       Low spender             17      735727        household
11       Low spender             14      700884        breakfast
12       Low spender             12      698706     meat seafood
13       Low spender             11      445617    personal care
14       Low spender             18      423368           babies
15       Low spender              6      268414    international
16       Low spender              5      152264          alcohol
17       Low spender              8       97329             pets
18       Low spender             21       68900          missing
19       Low spender              2       36211            other
20       Low spender             10       34505             bulk

Key Insights

Top Departments are Similar:
Both groups buy most from Produce and Dairy & Eggs – staples for everyone.

Volume Difference:
Low spenders purchase huge volumes of cheap staples (e.g., Snacks, Frozen, Pantry).
High spenders buy smaller quantities overall but spread spending across premium areas like Meat & Seafood and Personal Care.

Category Focus:
High spenders lean more toward fresh, perishable and premium items (Produce, Dairy, Meat & Seafood).
Low spenders heavily buy pantry, snacks, and frozen items — more cost-conscious choices.

Opportunities for Marketing:
Upsell premium deli or personal care products to low spenders.
Cross-sell household & specialty items to high spenders (they already buy premium foods).

High spenders behave like “quality seekers,” buying a variety of fresh and premium items. Low spenders are “volume shoppers,” focusing on essentials and budget-friendly categories. 

In [47]:
top_products_by_spender = (
    df_ords_prods_merge
    .groupby(['avg_spending_flag', 'product_name'])['product_id']
    .count()
    .reset_index()
    .sort_values(['avg_spending_flag', 'product_id'], ascending=[True, False])
)

In [48]:
top_products_by_spender

avg_spending_flag                                       product_name  \
841        High spender                                             Banana   
794        High spender                             Bag of Organic Bananas   
8186       High spender                                 Organic Whole Milk   
6887       High spender                               Organic Baby Spinach   
5332       High spender                                        Large Lemon   
...                 ...                                                ...   
60033       Low spender  Vitality Chicken Breasts with Flaxseed & Vitam...   
60089       Low spender                                  Vitamin D Gummies   
61219       Low spender                   Wicked Good No-Fry Wing Marinade   
61611       Low spender                               Yellow Fish Breading   
61896       Low spender           Zingz Queso Fundido Baked Snack Crackers   

       product_id  
841          2511  
794          1552  
8186         1348  
6887         1032  
5332          916  
...           ...  
60033           1  
60089           1  
61219           1  
61611           1  
61896           1  

[61989 rows x 3 columns]

In [49]:
# High spenders
high_spender_top_products = (
    top_products_by_spender[top_products_by_spender['avg_spending_flag'] == 'High spender']
    .sort_values('product_id', ascending=False)
)

In [50]:
high_spender_top_products.sort_values('product_id', ascending=False).head(20)

avg_spending_flag                                    product_name  \
841        High spender                                          Banana   
794        High spender                          Bag of Organic Bananas   
8186       High spender                              Organic Whole Milk   
6887       High spender                            Organic Baby Spinach   
5332       High spender                                     Large Lemon   
5575       High spender                                           Limes   
2760       High spender                                  Cucumber Kirby   
674        High spender                                       Asparagus   
12270      High spender                               Zero Calorie Cola   
9534       High spender                             Reduced Fat 2% Milk   
7994       High spender                            Organic Strawberries   
11602      High spender  Unsweetened Original Almond Breeze Almond Milk   
12076      High spender                                      Whole Milk   
6868       High spender                                 Organic Avocado   
4560       High spender                                     Half & Half   
10246      High spender                                      Smartwater   
10319      High spender                                            Soda   
9262       High spender                               Pure Irish Butter   
5995       High spender                           Michigan Organic Kale   
7          High spender                        0% Greek Strained Yogurt   

       product_id  
841          2511  
794          1552  
8186         1348  
6887         1032  
5332          916  
5575          801  
2760          643  
674           629  
12270         611  
9534          605  
7994          590  
11602         587  
12076         557  
6868          545  
4560          504  
10246         501  
10319         488  
9262          430  
5995          406  
7             390

In [51]:
# Low spenders
low_spender_top_products = (
    top_products_by_spender[top_products_by_spender['avg_spending_flag'] == 'Low spender']
    .sort_values('product_id', ascending=False)
)

In [52]:
low_spender_top_products.sort_values('product_id', ascending=False) 

avg_spending_flag                                       product_name  \
16003       Low spender                                             Banana   
15798       Low spender                             Bag of Organic Bananas   
44240       Low spender                               Organic Strawberries   
41160       Low spender                               Organic Baby Spinach   
42617       Low spender                               Organic Hass Avocado   
...                 ...                                                ...   
60033       Low spender  Vitality Chicken Breasts with Flaxseed & Vitam...   
60089       Low spender                                  Vitamin D Gummies   
61219       Low spender                   Wicked Good No-Fry Wing Marinade   
61611       Low spender                               Yellow Fish Breading   
61896       Low spender           Zingz Queso Fundido Baked Snack Crackers   

       product_id  
16003      470054  
15798      377898  
44240      264093  
41160      240889  
42617      213303  
...           ...  
60033           1  
60089           1  
61219           1  
61611           1  
61896           1  

[49661 rows x 3 columns]

In [53]:
low_spender_top_products.head(50)

avg_spending_flag                          product_name  product_id
16003       Low spender                                Banana      470054
15798       Low spender                Bag of Organic Bananas      377898
44240       Low spender                  Organic Strawberries      264093
41160       Low spender                  Organic Baby Spinach      240889
42617       Low spender                  Organic Hass Avocado      213303
41124       Low spender                       Organic Avocado      176270
34734       Low spender                           Large Lemon      151741
55219       Low spender                          Strawberries      142688
35741       Low spender                                 Limes      139826
43683       Low spender                   Organic Raspberries      136922
44797       Low spender                    Organic Whole Milk      136557
44884       Low spender                  Organic Yellow Onion      113312
42320       Low spender                        Organic Garlic      109560
44924       Low spender                      Organic Zucchini      104629
41328       Low spender                   Organic Blueberries       99757
23955       Low spender                        Cucumber Kirby       96672
42300       Low spender                    Organic Fuji Apple       89489
42897       Low spender                         Organic Lemon       87599
14955       Low spender              Apple Honeycrisp Organic       84686
42459       Low spender                Organic Grape Tomatoes       84069
52019       Low spender                   Seedless Red Grapes       82555
41923       Low spender                      Organic Cucumber       80029
32706       Low spender                      Honeycrisp Apple       79631
41136       Low spender                  Organic Baby Carrots       76529
42595       Low spender                   Organic Half & Half       76294
54207       Low spender            Sparkling Water Grapefruit       75771
42880       Low spender  Organic Large Extra Fancy Fuji Apple       74973
61623       Low spender                         Yellow Onions       73032
41131       Low spender                  Organic Baby Arugula       72702
42310       Low spender                   Organic Gala Apples       72686
19126       Low spender                               Carrots       72486
28055       Low spender                     Fresh Cauliflower       71468
45345       Low spender                       Original Hummus       71213
41707       Low spender                      Organic Cilantro       69411
31384       Low spender                           Half & Half       68713
44026       Low spender            Organic Small Bunch Celery       67942
37551       Low spender                 Michigan Organic Kale       67043
43773       Low spender                     Organic Red Onion       66850
15339       Low spender                             Asparagus       66654
44455       Low spender                Organic Tomato Cluster       64020
41301       Low spender                  Organic Blackberries       61534
12787       Low spender                100% Whole Wheat Bread       60684
42761       Low spender         Organic Italian Parsley Bunch       60524
44829       Low spender           Organic Whole String Cheese       59591
42451       Low spender            Organic Granny Smith Apple       58749
44566       Low spender       Organic Unsweetened Almond Milk       57849
49455       Low spender                           Raspberries       57537
43749       Low spender               Organic Red Bell Pepper       57394
54736       Low spender                          Spring Water       55984
17307       Low spender                           Blueberries       55665

#### Top Product Insights – High vs. Low Spenders

High Spenders:
Gravitate toward premium organics: Organic Whole Milk, Organic Baby Spinach, Organic Strawberries, Organic Avocado, Almond Milk.
Also purchase specialty dairy (Greek Yogurt, Half & Half, Irish Butter) and branded drinks (Smartwater, Zero Calorie Cola).
Their baskets show a focus on quality, variety, and health-conscious choices.

Low Spenders:
Buy similar staples (bananas, spinach, strawberries, avocados), but they make these purchases across more orders and more customers, making these items the most common across their group.
Their focus is on essential, affordable produce and basics rather than premium or niche items.

#### Comparison & Marketing Takeaway:

Both groups buy the same core fresh products (bananas, spinach, strawberries).
High Spenders lean toward premium and specialty versions, while Low Spenders stick with staples and essentials.

Marketing can:
Offer premium cross-sells and niche products to High Spenders.
Use value bundles, family packs, or discounts to engage Low Spenders.
Bottom line: High Spenders are quality-focused; Low Spenders are budget- and staple-focused — and campaigns should reflect those differences.

In [54]:
df_ords_prods_merge.shape

(32404859, 25)

In [55]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... price_range_loc     Busiest day        Busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regularly busy day   
1          0  ...       Mid-range  Regularly busy  Regularly busy day   
2          0  ...       Low-range  Regularly busy  Regularly busy day   
3          0  ...       Low-range  Regularly busy  Regularly busy day   
4          0  ...       Low-range  Regularly busy  Regularly busy day   

   busiest_period_of_day max_order  loyalty_flag avg_price avg_spending_flag  \
0         Average orders        10  New customer  6.367797       Low spender   
1         Average orders        10  New customer  6.367797       Low spender   
2         Average orders        10  New customer  6.367797       Low spender   
3         Average orders        10  New customer  6.367797       Low spender   
4         Average orders        10  New customer  6.367797       Low spender   

  user_median_order_frequency   order_frequency_flag  
0                        20.5  Non-frequent customer  
1                        20.5  Non-frequent customer  
2                        20.5  Non-frequent customer  
3                        20.5  Non-frequent customer  
4                        20.5  Non-frequent customer  

[5 rows x 25 columns]

# 06 Exporting Data

In [58]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', '4.8_ords_prods_merge_final.pkl'))